File: src/2022/day04/day04.py
"""After completing this in C++ I wanted to make a set based approach too"""

import re
import sys

from libs.aoc_runner.python import aoc_runner


def sol1(list_of_pairs: list[tuple[set[int], set[int]]]) -> int:
    """Counts occurences by checking if the sets are subsets to one another"""
    count = 0
    for elf_1, elf_2 in list_of_pairs:
        if elf_1.issubset(elf_2) or elf_2.issubset(elf_1):
            count += 1

    return count


def sol2(list_of_pairs: list[tuple[set[int], set[int]]]) -> int:
    """Checks if the sets intersects at any points"""
    count = 0
    for elf_1, elf_2 in list_of_pairs:
        if len(elf_1 & elf_2) > 0:
            count += 1

    return count


def read_and_parse(filename: str) -> list[tuple[set[int], set[int]]]:
    """Reads the file line by line and parses the pattern
    d_1-d_2,d_3-d_4
    into 2 sets defined by {d_1..d_2}, {d_3..d_4} and returns a list of these pairs
    """

    with open(filename, "r", encoding="utf-8") as open_file:
        lines = open_file.readlines()
        list_of_pairs = []
        for line in lines:
            d_1, d_2, d_3, d_4 = map(int, re.findall(r"\d+", line))

            elf_1 = set(range(d_1, d_2 + 1))
            elf_2 = set(range(d_3, d_4 + 1))

            list_of_pairs.append((elf_1, elf_2))
        return list_of_pairs


def samples():
    sample = read_and_parse("day04-sample.input")
    assert sol1(sample) == 2
    assert sol2(sample) == 4


def solve_1(x):
    res = sol1(x)
    print(f"part 1: {res}")
    assert res == 466


def solve_2(x):
    res = sol2(x)
    print(f"part 2: {res}")
    assert res == 865


if __name__ == "__main__":
    inp = read_and_parse("day04.input")
    sys.exit(
        aoc_runner.aoc_runner(
            samples,
            solve_1,
            solve_2,
            inp,
        )
    )

File: src/2022/day04/day04.cpp
#include <cassert>
#include <iostream>
#include <string>
#include <tuple>
#include <vector>

#include "aoc_runner.h"
#include "aoc_utils.h"

std::vector<std::tuple<int, int, int, int>> read_and_parse_data(std::vector<std::string> const& input) {
    std::vector<std::tuple<int, int, int, int>> pairs;
    for (auto const& line : input) {
        int elf_1_start, elf_1_end;
        int elf_2_start, elf_2_end;
        sscanf(line.c_str(), "%d-%d,%d-%d", &elf_1_start, &elf_1_end, &elf_2_start, &elf_2_end);
        pairs.emplace_back(std::make_tuple(elf_1_start, elf_1_end, elf_2_start, elf_2_end));
    }
    return pairs;
}

int solve_1(std::vector<std::tuple<int, int, int, int>> input) {
    int count = 0;
    for (auto& it : input) {
        auto [a, b, c, d] = it;
        if ((a <= c && d <= b) || (c <= a && b <= d)) {
            ++count;
        }
    }
    return count;
}

int solve_2(std::vector<std::tuple<int, int, int, int>> input) {
    int count = 0;
    for (auto& it : input) {
        auto [a, b, c, d] = it;
        if (!((b < c) || (d < a))) {
            ++count;
        }
    }
    return count;
}

void samples() {
    auto sample = aoc::utils::read_input(AOC_SAMPLE_INPUT);
    auto tmp = read_and_parse_data(sample);
    assert(solve_1(tmp) == 2);
    assert(solve_2(tmp) == 4);
}

int main(int argc, char** argv) {
    auto input = aoc::utils::read_input(AOC_INPUT);

    auto solve_1_wrapper = [](std::vector<std::string> const& inp) -> void {
        auto tmp = read_and_parse_data(inp);
        auto part1 = solve_1(tmp);
        std::cout << "part 1: " << part1 << std::endl;
    };
    auto solve_2_wrapper = [](std::vector<std::string> const& inp) -> void {
        auto tmp = read_and_parse_data(inp);
        auto part2 = solve_2(tmp);
        std::cout << "part 2: " << part2 << std::endl;
    };

    return aoc::run(argc, argv, samples, solve_1_wrapper, solve_2_wrapper, input);
}
